Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean

```diff
@@ -32,48 +33,58 @@ generalized to fully semibilinear forms.
 
 ## Tags
 
-sesquilinear_form, matrix, basis
+Sesquilinear form, Sesquilinear map, matrix, basis
 
 -/
 
 
-variable {R R₁ R₂ M M₁ M₂ M₁' M₂' n m n' m' ι : Type*}
+variable {R R₁ R₂ M M₁ M₂ M₁' M₂' N₂ n m n' m' ι : Type*}
 
 open BigOperators
 
 open Finset LinearMap Matrix
 
 open Matrix
 
+open scoped RightActions
+
 section AuxToLinearMap
 
-variable [CommSemiring R] [Semiring R₁] [Semiring R₂]
+variable [CommSemiring R] [Semiring R₁] [Semiring R₂] [AddCommMonoid N₂] [Module R N₂]
+  [Module Rᵐᵒᵖ N₂] [IsCentralScalar R N₂]
 
 variable [Fintype n] [Fintype m]
 
 variable (σ₁ : R₁ →+* R) (σ₂ : R₂ →+* R)
 
-/-- The map from `Matrix n n R` to bilinear forms on `n → R`.
+/-- The map from `Matrix n n R` to bilinear maps on `n → R`.
 
 This is an auxiliary definition for the equivalence `Matrix.toLinearMap₂'`. -/
-def Matrix.toLinearMap₂'Aux (f : Matrix n m R) : (n → R₁) →ₛₗ[σ₁] (m → R₂) →ₛₗ[σ₂] R :=
+def Matrix.toLinearMap₂'Aux (f : Matrix n m N₂) : (n → R₁) →ₛₗ[σ₁] (m → R₂) →ₛₗ[σ₂] N₂ :=
```

Comment:
This seems pretty fishy to me. I think the correct generality is
```lean
variable [Semiring R₁] [Semiring R₁'] [Semiring R₂] [Semiring R₂'] [AddCommMonoid M]
variable [Module R₁' M] [Module R₂' M] [SMulCommClass R₂' R₁' M]
variable (σ₁ : R₁ →+* R₁') (σ₂ : R₂ →+* R₂')

variable [Fintype n] [Fintype m]

/-- The map from `Matrix n n R` to bilinear maps on `n → R`.

This is an auxiliary definition for the equivalence `Matrix.toLinearMap₂'`. -/
def Matrix.toLinearMap₂'Aux (f : Matrix n m M) : (n → R₁) →ₛₗ[σ₁] (m → R₂) →ₛₗ[σ₂] M :=
  -- porting note: we don't seem to have `∑ i j` as valid notation yet
  mk₂'ₛₗ σ₁ σ₂ (fun (v : n → R₁) (w : m → R₂) => ∑ i, ∑ j, σ₂ (w j) • σ₁ (v i) • f i j)
    (fun _ _ _ => by simp only [Pi.add_apply, map_add, smul_add, sum_add_distrib, add_smul])
    (fun c v w => by
      simp only [Pi.smul_apply, smul_sum, smul_eq_mul, σ₁.map_mul, ← smul_comm _ (σ₁ c),
        MulAction.mul_smul])
    (fun _ _ _ => by simp only [Pi.add_apply, map_add, add_smul, smul_add, sum_add_distrib])
    (fun _ v w => by
      simp only [Pi.smul_apply, smul_eq_mul, _root_.map_mul, MulAction.mul_smul, smul_sum])
```
as this should then work for `(f : Matrix n m ℍ) : (n → ℍ) →ₗ[ℍ] (m → ℍ) →ₛₗ[starRingEquiv] ℍ`

Reply:
More precisely, that definition works with:
```
example {R : Type*} [Ring R] [StarRing R] (f : Matrix n m R) :
    (n → R) →ₗ[R] (m → Rᵐᵒᵖ) →ₗ[Rᵐᵒᵖ] R :=
  Matrix.toLinearMap₂'Aux _ _ f
example {R : Type*} [Ring R] [StarRing R] (f : Matrix n m R) :
    (n → R) →ₗ[R] (m → R) →ₛₗ[(starRingEquiv (R := R)).toRingHom] R :=
  Matrix.toLinearMap₂'Aux _ _ f
```

Reply:
@eric-wieser would you then have:
```
variable [AddCommMonoid M₁] [Module R₁ M₁] [AddCommMonoid M₂] [Module R₂ M₂] [AddCommMonoid N₂]
  [Module R N₂] [SMulCommClass R R N₂]

variable {σ₁ : R₁ →+* R} {σ₂ : R₂ →+* R}

/-- The linear map from sesquilinear maps to `Matrix n m R` given an `n`-indexed basis for `M₁`
and an `m`-indexed basis for `M₂`.

This is an auxiliary definition for the equivalence `Matrix.toLinearMapₛₗ₂'`. -/
def LinearMap.toMatrix₂Aux (b₁ : n → M₁) (b₂ : m → M₂) :
    (M₁ →ₛₗ[σ₁] M₂ →ₛₗ[σ₂] N₂) →ₗ[R] Matrix n m N₂ where
  toFun f := of fun i j => f (b₁ i) (b₂ j)
  map_add' _f _g := rfl
  map_smul' _f _g := rfl
```
etc?

Reply:
What's the assumption on `R` in that example? I think probably you're missing another ring or two there; the fact that conversion to a matrix is itself a linear operation is independent of the linear operation itself, so I am suspicious of the `→ₗ[R]`.

Reply:
So, something like:
```
variable [Semiring R] [Semiring R₁] [Semiring S₁] [Semiring R₂] [Semiring S₂]

variable [AddCommMonoid M₁] [Module R₁ M₁] [AddCommMonoid M₂] [Module R₂ M₂] [AddCommMonoid N₂]
  [Module R N₂] [Module S₁ N₂] [Module S₂ N₂] [SMulCommClass S₁ R N₂] [SMulCommClass S₂ R N₂] 
  [SMulCommClass S₂ S₁ N₂] 

variable {σ₁ : R₁ →+* S₁} {σ₂ : R₂ →+* S₂}

/-- The linear map from sesquilinear maps to `Matrix n m N₂` given an `n`-indexed basis for `M₁`
and an `m`-indexed basis for `M₂`.

This is an auxiliary definition for the equivalence `Matrix.toLinearMapₛₗ₂'`. -/
def LinearMap.toMatrix₂Aux (b₁ : n → M₁) (b₂ : m → M₂) :
    (M₁ →ₛₗ[σ₁] M₂ →ₛₗ[σ₂] N₂) →ₗ[R] Matrix n m N₂ where
  toFun f := of fun i j => f (b₁ i) (b₂ j)
  map_add' _f _g := rfl
  map_smul' _f _g := rfl
```
?

Reply:
Yes, that looks sensible

Reply:
I'd got it in my head that we needed to be wary of using `SMulCommClass` too much for performance reasons, but maybe I got hold of the wrong end of the stick?

Reply:
That's almost certainly true, but also we don't want to do it at the expense of generalizing in the wrong direction.

Reply:
@eric-wieser I'm afraid I don't quite have the brain power to see what `LinearMap.toMatrix₂` should become? I'm thinking something like:
```
variable [CommSemiring R] [CommSemiring R₁] [CommSemiring R₂]

variable [AddCommMonoid M₁] [Module R₁ M₁] [AddCommMonoid M₂] [Module R₂ M₂] [AddCommMonoid N₂]
  [Module R N₂] [Module R₁ N₂] [Module R₂ N₂] [SMulCommClass R₂ R N₂] [SMulCommClass R₁ R N₂]
  [SMulCommClass R₂ R₁ N₂] --[IsCentralScalar R N₂]

variable [DecidableEq n] [Fintype n]

variable [DecidableEq m] [Fintype m]

variable (b₁ : Basis n R₁ M₁) (b₂ : Basis m R₂ M₂)

/-- `LinearMap.toMatrix₂ b₁ b₂` is the equivalence between `R`-bilinear maps on `M` and
`n`-by-`m` matrices with entries in `R`, if `b₁` and `b₂` are `R`-bases for `M₁` and `M₂`,
respectively. -/
noncomputable def LinearMap.toMatrix₂ : (M₁ →ₗ[R₁] M₂ →ₗ[R₂] N₂) ≃ₗ[R] Matrix n m N₂ := sorry
```
We have:
```
#check (b₂.equivFun.arrowCongr (M₂₁ := N₂) (M₂₂ := N₂) (LinearEquiv.refl R₂ N₂) : (M₂ →ₗ[R₂] N₂) ≃ₗ[R₂] (m → R₂) →ₗ[R₂] N₂)
```
but then I'm not sure what to do next assuming `R₁` and `R₂` are not equal?

Reply:
I think the problem is that I don't see a useful way to generalise `LinearEquiv.arrowCongr`.

Reply:
Maybe something like:
```
def arrowCongr' {R M₁ M₂ M₂₁ M₂₂ : Sort _} [CommSemiring R] [AddCommMonoid M₁] [AddCommMonoid M₂]
    [AddCommMonoid M₂₁] [AddCommMonoid M₂₂]
    [Module R₁ M₁] [Module R₁ M₂] [Module R₁ M₂₁] [Module R₁ M₂₂]
    [Module R M₂₁] [Module R M₂₂]
    [SMulCommClass R₁ R M₂₁]
    [SMulCommClass R₁ R M₂₂]
    (e₁ : M₁ ≃ₗ[R₁] M₂) (e₂ : M₂₁ ≃ₗ[R₁] M₂₂) (h : IsLinearMap R e₂): (M₁ →ₗ[R₁] M₂₁) ≃ₗ[R] M₂ →ₗ[R₁] M₂₂
```

Reply:
This then appears to work:
```
noncomputable def LinearMap.toMatrix₂ : (M₁ →ₗ[R₁] M₂ →ₗ[R₂] N₂) ≃ₗ[R] Matrix n m N₂ :=
  (b₁.equivFun.arrowCongr' (b₂.equivFun.arrowCongr' (LinearEquiv.refl R₂ N₂)
    ({ map_add := fun _ ↦ congrFun rfl, map_smul := fun _ ↦ congrFun rfl })) (
      { map_add := fun x ↦ congrFun rfl, map_smul := fun c ↦ congrFun (by exact rfl) })).trans
    LinearMap.toMatrix₂'
```

`(h : IsLinearMap R e₂)` should probably just be the scalar multiplication

Reply:
Okay, I think I've fixed this now. Thanks!

Reply:
@eric-wieser do you have an alternative suggestion for how this should look?